lib/kube/token/validator.go

```diff
@@ -85,49 +83,84 @@ func (c *ValidationResult) JoinAuditAttributes() (map[string]interface{}, error)
+	clusterAudiences []string
 }
 
-// getClient allows the lazy initialisation of the Kubernetes client
-func (v *TokenReviewValidator) getClient() (kubernetes.Interface, error) {
+// getClient allows the lazy initialisation of the Kubernetes client and clusterAudiences
+func (v *TokenReviewValidator) getClient(_ context.Context) (kubernetes.Interface, []string, error) {
 	v.mu.Lock()
 	defer v.mu.Unlock()
 	if v.client != nil {
-		return v.client, nil
+		return v.client, v.clusterAudiences, nil
 	}
 
 	config, err := rest.InClusterConfig()
 	if err != nil {
-		return nil, trace.WrapWithMessage(err, "failed to initialize in-cluster Kubernetes config")
+		return nil, nil, trace.WrapWithMessage(err, "failed to initialize in-cluster Kubernetes config")
 	}
 	client, err := kubernetes.NewForConfig(config)
 	if err != nil {
-		return nil, trace.WrapWithMessage(err, "failed to initialize in-cluster Kubernetes client")
+		return nil, nil, trace.WrapWithMessage(err, "failed to initialize in-cluster Kubernetes client")
+	}
+
+	// We do a self-review to recover the default cluster audience
+	audiences, err := getTokenAudiences(config.BearerToken)
+	if err != nil {
+		return nil, nil, trace.Wrap(err, "doing a self-review")
 	}
 
 	v.client = client
-	return client, nil
+	v.clusterAudiences = audiences
+	return client, audiences, nil
+}
+
+// getTokenAudiences extracts the audience from
+func getTokenAudiences(token string) ([]string, error) {
+	jwt, err := josejwt.ParseSigned(token)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	claims := &ServiceAccountClaims{}
+	err = jwt.UnsafeClaimsWithoutVerification(claims)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return claims.Audience, nil
 }
 
 // Validate uses the Kubernetes TokenReview API to validate a token and return its UserInfo
 func (v *TokenReviewValidator) Validate(ctx context.Context, token, clusterName string) (*ValidationResult, error) {
-	client, err := v.getClient()
+	client, audiences, err := v.getClient(ctx)
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
 
 	review := &v1.TokenReview{
 		Spec: v1.TokenReviewSpec{
 			Token: token,
-			// In-cluster used to only allow tokens with the kubernetes audience
-			// But people kept confusing it with JWKS and set the cluster name
-			// as the audience/. To avoid his common footgun we now allow tokens
-			// whose audience is the teleport cluster name.
-			Audiences: []string{kubernetesAudience, clusterName},
 		},
 	}
+
+	// In-cluster used to only allow tokens with the kubernetes audience but people
+	// kept confusing it with the JWKS kube join method and set the cluster name
+	// as the audience. To avoid his common footgun we now allow tokens whose
+	// audience is the teleport cluster name.
+	//
+	// We do this only if the Kubernetes cluster supports audiences.
+	// Earlier Kube versions don't have audience
+	// support, in this case, we just do a regular token review.
+	if audiences != nil {
```

Comment:
I don't think `go-jose` makes promises wrt empty vs missing `aud` (unless I'm missing something).

```suggestion
	if len(audiences) < 1 {
```